File: iosApp/iosApp/CrashReporterBridge.swift
import Foundation
import FirebaseCore
import FirebaseCrashlytics
import Shared

/// shared (Kotlin/Native) 의 `CrashReporterBridge` handler 를 세팅해 non-fatal 예외를 Crashlytics 로 전송.
///
/// - `iOSApp.init` 에서 `configure()` 호출 (FirebaseApp.configure 이후에).
/// - shared 는 `CrashReporter.recordException(...)` 만 호출하고, 실제 Crashlytics 호출은 여기서 담당.
/// - TestFlight/Release 빌드에서 카카오 로그인 실패 원인을 원격에서 확인하기 위한 용도.
enum CrashReporterInstaller {

    static func configure() {
        CrashReporterBridge.shared.logHandler = { message in
            Crashlytics.crashlytics().log(message)
        }
        CrashReporterBridge.shared.recordHandler = { message, exceptionName, reason in
            if let message = message, !message.isEmpty {
                Crashlytics.crashlytics().log(message)
            }
            let domain = exceptionName.isEmpty ? "KotlinThrowable" : exceptionName
            let userInfo: [String: Any] = [
                NSLocalizedDescriptionKey: reason ?? "unknown",
                "breadcrumb": message ?? "",
            ]
            let error = NSError(domain: domain, code: 0, userInfo: userInfo)
            Crashlytics.crashlytics().record(error: error)
        }
    }

    /// Swift 쪽 (iOSApp.swift KakaoLoginBridge.handler) 에서 잡은 error 를 직접 Crashlytics 로 보내는 헬퍼.
    /// shared 의 KakaoLoginResult.Failure 로 매핑되기 전에 원본 NSError 를 그대로 남겨 원인 파악에 유리.
    static func recordKakaoLoginError(_ error: Error, breadcrumb: String) {
        Crashlytics.crashlytics().log("[Kakao] " + breadcrumb)
        Crashlytics.crashlytics().record(error: error)
    }
}

File: shared/src/iosMain/kotlin/com/hyunjine/linker/platform/CrashReporterBridge.kt
package com.hyunjine.linker.platform

/**
 * iOS Kotlin/Native ↔ Swift 브리지. FirebaseCrashlytics 는 Swift 로 배포되어 Kotlin/Native 에서 직접
 * 참조가 어려워, Swift 쪽에서 [Crashlytics] 호출을 수행하고 shared 는 이 브리지로만 위임한다.
 *
 * 사용 (iOSApp.swift):
 * ```swift
 * CrashReporterBridge.shared.recordHandler = { message, name, reason in
 *     let userInfo: [String: Any] = [ NSLocalizedDescriptionKey: reason ?? "unknown", ... ]
 *     Crashlytics.crashlytics().record(error: NSError(domain: name, code: 0, userInfo: userInfo))
 * }
 * CrashReporterBridge.shared.logHandler = { message in
 *     Crashlytics.crashlytics().log(message)
 * }
 * ```
 *
 * handler 미세팅 시 recordException 은 조용히 무시 (Swift 초기화 전 호출 방어).
 */
object CrashReporterBridge {

    /**
     * Swift 에서 세팅. shared 가 non-fatal 예외를 기록하고 싶을 때 호출한다.
     * - [message]: 브레드크럼 메시지 (nullable)
     * - [exceptionName]: Throwable 의 클래스 심플명 (`error:` 의 NSError.domain 으로 사용)
     * - [reason]: Throwable.message (사용자에게는 노출하지 않는 개발자용 사유)
     */
    var recordHandler: ((message: String?, exceptionName: String, reason: String?) -> Unit)? = null

    /** Swift 에서 세팅. 다음 record 리포트에 첨부될 브레드크럼 로그. */
    var logHandler: ((message: String) -> Unit)? = null
}
